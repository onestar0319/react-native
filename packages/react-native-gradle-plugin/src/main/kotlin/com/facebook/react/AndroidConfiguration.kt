/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react

import com.android.build.gradle.BaseExtension
import org.gradle.api.Project

fun Project.configureDevPorts(androidExt: BaseExtension) {
  val devServerPort = project.properties["reactNativeDevServerPort"]?.toString() ?: "8081"
  val inspectorProxyPort = project.properties["reactNativeInspectorProxyPort"]?.toString() ?: devServerPort

  androidExt.buildTypes.all {
    resValue("integer", "react_native_dev_server_port", devServerPort)
    resValue("integer", "react_native_inspector_proxy_port", inspectorProxyPort)
  }
}
