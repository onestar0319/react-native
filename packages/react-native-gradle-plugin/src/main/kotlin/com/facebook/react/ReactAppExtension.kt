/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react

import com.android.build.gradle.api.BaseVariant
import org.apache.tools.ant.taskdefs.condition.Os
import org.gradle.api.Project
import java.io.File

open class ReactAppExtension(private val project: Project) {
  var composeSourceMapsPath: String = "node_modules/react-native/scripts/compose-source-maps.js"
  var bundleAssetName: String = "index.android.bundle"
  var entryFile: File? = null
  var bundleCommand: String = "bundle"
  var reactRoot: File = File(project.projectDir, "../../")
  var inputExcludes: List<String> = listOf("android/**", "ios/**")
  var bundleConfig: String? = null
  var enableVmCleanup: Boolean = true
  var hermesCommand: String = "node_modules/hermes-engine/%OS-BIN%/hermesc"
  var cliPath: String? = null
  var nodeExecutableAndArgs: List<String> = listOf("node")
  var enableHermes: Boolean = false
  var enableHermesForVariant: (BaseVariant) -> Boolean = { enableHermes }
  var devDisabledInVariants: List<String> = emptyList()
  // todo maybe lambda as for hermes?
  var bundleIn: Map<String, Boolean> = emptyMap()
  var extraPackagerArgs: List<String> = emptyList()
  var hermesFlagsDebug: List<String> = emptyList()
  var hermesFlagsRelease: List<String> = listOf("-O", "-output-source-map")
  var resourcesDir: Map<String, File> = emptyMap()
  var jsBundleDir: Map<String, File> = emptyMap()

  internal val detectedEntryFile: File
    get() = detectEntryFile(entryFile = entryFile, reactRoot = reactRoot)

  internal val detectedCliPath: String
    get() = detectCliPath(
      projectDir = project.projectDir,
      reactRoot = reactRoot,
      preconfuredCliPath = cliPath
    )

  internal val osAwareHermesCommand: String
    get() = getOSAwareHermesCommand(hermesCommand)

  private fun detectEntryFile(entryFile: File?, reactRoot: File): File = when {
    System.getenv("ENTRY_FILE") != null -> File(System.getenv("ENTRY_FILE"))
    entryFile != null -> entryFile
    File(reactRoot, "index.android.js").exists() -> File(reactRoot, "index.android.js")
    else -> File(reactRoot, "index.android.js")
  }

  private fun detectCliPath(projectDir: File, reactRoot: File, preconfuredCliPath: String?): String {
    // 1. preconfigured path
    if (preconfuredCliPath != null) return preconfuredCliPath

    // 2. node module path
    val nodeProcess = Runtime.getRuntime().exec(
      arrayOf("node", "-e", "console.log(require('react-native/cli').bin);"),
      emptyArray(),
      projectDir
    )

    val nodeProcessOutput = nodeProcess.inputStream.use {
      it.bufferedReader().readText().trim()
    }

    if (nodeProcessOutput.isNotEmpty()) {
      return nodeProcessOutput
    }

    // 3. cli.js in the root folder
    val rootCliJs = File(reactRoot, "node_modules/react-native/cli.js")
    if (rootCliJs.exists()) {
      return rootCliJs.absolutePath
    }

    error("Couldn't determine CLI location. " +
      "Please set `project.react.cliPath` to the path of the react-native cli.js")
  }

  // Make sure not to inspect the Hermes config unless we need it,
  // to avoid breaking any JSC-only setups.
  private fun getOSAwareHermesCommand(hermesCommand: String): String {
    // If the project specifies a Hermes command, don't second guess it.
    if (!hermesCommand.contains("%OS-BIN%")) {
      return hermesCommand
    }

    // Execution on Windows fails with / as separator
    return hermesCommand
      .replace("%OS-BIN%", getHermesOSBin())
      .replace('/', File.separatorChar)
  }

  private fun getHermesOSBin(): String {
    if (Os.isFamily(Os.FAMILY_WINDOWS)) return "win64-bin"
    if (Os.isFamily(Os.FAMILY_MAC)) return "osx-bin"
    if (Os.isOs(null, "linux", "amd64", null)) return "linux64-bin"
    error("OS not recognized. Please set project.react.hermesCommand " +
      "to the path of a working Hermes compiler.")
  }
}
