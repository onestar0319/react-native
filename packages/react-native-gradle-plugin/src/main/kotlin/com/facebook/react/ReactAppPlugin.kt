/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react

import com.android.build.gradle.AppExtension
import com.android.build.gradle.BaseExtension
import com.android.build.gradle.LibraryExtension
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.kotlin.dsl.create
import org.gradle.kotlin.dsl.getByType

class ReactAppPlugin : Plugin<Project> {
  override fun apply(project: Project) {
    val config = project.extensions.create<ReactAppExtension>("reactApp", project)

    project.afterEvaluate {
      val androidConfiguration = extensions.getByType<BaseExtension>()
      configureDevPorts(androidConfiguration)

      val isAndroidLibrary = plugins.hasPlugin("com.android.library")
      val variants = if (isAndroidLibrary) {
        extensions.getByType<LibraryExtension>().libraryVariants
      } else {
        extensions.getByType<AppExtension>().applicationVariants
      }
      variants.all {
        configureReactTasks(
          variant = this,
          config = config
        )
      }
    }
  }
}
