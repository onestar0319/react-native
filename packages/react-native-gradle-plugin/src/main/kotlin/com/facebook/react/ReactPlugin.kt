/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react

import com.android.build.api.variant.AndroidComponentsExtension
import com.android.build.gradle.AppExtension
import com.android.build.gradle.internal.tasks.factory.dependsOn
import com.facebook.react.tasks.BuildCodegenCLITask
import com.facebook.react.tasks.GenerateCodegenArtifactsTask
import com.facebook.react.tasks.GenerateCodegenSchemaTask
import com.facebook.react.utils.AgpConfiguratorUtils.configureBuildConfigFields
import com.facebook.react.utils.AgpConfiguratorUtils.configureDevPorts
import com.facebook.react.utils.BackwardCompatUtils.configureBackwardCompatibilityReactMap
import com.facebook.react.utils.DependencyUtils.configureDependencies
import com.facebook.react.utils.DependencyUtils.configureRepositories
import com.facebook.react.utils.DependencyUtils.readVersionString
import com.facebook.react.utils.JsonUtils
import com.facebook.react.utils.NdkConfiguratorUtils.configureReactNativeNdk
import com.facebook.react.utils.ProjectUtils.needsCodegenFromPackageJson
import com.facebook.react.utils.findPackageJsonFile
import java.io.File
import kotlin.system.exitProcess
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.Task
import org.gradle.internal.jvm.Jvm

class ReactPlugin : Plugin<Project> {
  override fun apply(project: Project) {
    checkJvmVersion(project)
    val extension = project.extensions.create("react", ReactExtension::class.java, project)

    // App Only Configuration
    project.pluginManager.withPlugin("com.android.application") {
      project.afterEvaluate {
        val reactNativeDir = extension.reactNativeDir.get().asFile
        val propertiesFile = File(reactNativeDir, "ReactAndroid/gradle.properties")
        val versionString = readVersionString(propertiesFile)
        configureDependencies(project, versionString)
        configureRepositories(project, reactNativeDir)
      }

      configureReactNativeNdk(project, extension)
      configureBuildConfigFields(project)
      configureDevPorts(project)
      configureBackwardCompatibilityReactMap(project)

      project.extensions.getByType(AndroidComponentsExtension::class.java).apply {
        onVariants(selector().all()) { variant ->
          project.configureReactTasks(variant = variant, config = extension)
        }
      }

      // This is a legacy AGP api. Needed as AGP 7.3 is not consuming generated resources correctly.
      // Can be removed as we bump to AGP 7.4 stable.
      // This registers the $buildDir/generated/res/react/<variant> folder as a
      // res folder to be consumed with the old AGP Apis which are not broken.
      project.extensions.getByType(AppExtension::class.java).apply {
        this.applicationVariants.all { variant ->
          val isDebuggableVariant =
              extension.debuggableVariants.get().any { it.equals(variant.name, ignoreCase = true) }
          val targetName = variant.name.replaceFirstChar { it.uppercase() }
          val bundleTaskName = "createBundle${targetName}JsAndAssets"
          if (!isDebuggableVariant) {
            variant.registerGeneratedResFolders(
                project.layout.buildDirectory.files("generated/res/react/${variant.name}"))
            variant.mergeResourcesProvider.get().dependsOn(bundleTaskName)
          }
        }
      }
      configureCodegen(project, extension, isLibrary = false)
    }

    // Library Only Configuration
    project.pluginManager.withPlugin("com.android.library") {
      configureCodegen(project, extension, isLibrary = true)
    }
  }

  private fun checkJvmVersion(project: Project) {
    val jvmVersion = Jvm.current()?.javaVersion?.majorVersion
    if ((jvmVersion?.toIntOrNull() ?: 0) <= 8) {
      project.logger.error(
          """

      ********************************************************************************

      ERROR: requires JDK11 or higher.
      Incompatible major version detected: '$jvmVersion'

      ********************************************************************************

      """
              .trimIndent())
      exitProcess(1)
    }
  }

  /**
   * A plugin to enable react-native-codegen in Gradle environment. See the Gradle API docs for more
   * information: https://docs.gradle.org/current/javadoc/org/gradle/api/Project.html
   */
  @Suppress("UnstableApiUsage")
  private fun configureCodegen(project: Project, extension: ReactExtension, isLibrary: Boolean) {
    // First, we set up the output dir for the codegen.
    val generatedSrcDir = File(project.buildDir, "generated/source/codegen")

    // We specify the default value (convention) for jsRootDir.
    // It's the root folder for apps (so ../../ from the Gradle project)
    // and the package folder for library (so ../ from the Gradle project)
    if (isLibrary) {
      extension.jsRootDir.convention(project.layout.projectDirectory.dir("../"))
    } else {
      extension.jsRootDir.convention(extension.root)
    }

    val buildCodegenTask =
        project.tasks.register("buildCodegenCLI", BuildCodegenCLITask::class.java) {
          it.codegenDir.set(extension.codegenDir)
          val bashWindowsHome = project.findProperty("REACT_WINDOWS_BASH") as String?
          it.bashWindowsHome.set(bashWindowsHome)

          // Please note that appNeedsCodegen is triggering a read of the package.json at
          // configuration time as we need to feed the onlyIf condition of this task.
          // Therefore, the appNeedsCodegen needs to be invoked inside this lambda.
          val needsCodegenFromPackageJson = project.needsCodegenFromPackageJson(extension)
          it.onlyIf { isLibrary || needsCodegenFromPackageJson }
        }

    // We create the task to produce schema from JS files.
    val generateCodegenSchemaTask =
        project.tasks.register(
            "generateCodegenSchemaFromJavaScript", GenerateCodegenSchemaTask::class.java) { it ->
              it.dependsOn(buildCodegenTask)
              it.nodeExecutableAndArgs.set(extension.nodeExecutableAndArgs)
              it.codegenDir.set(extension.codegenDir)
              it.generatedSrcDir.set(generatedSrcDir)

              // We're reading the package.json at configuration time to properly feed
              // the `jsRootDir` @Input property of this task & the onlyIf. Therefore, the
              // parsePackageJson should be invoked inside this lambda.
              val packageJson = findPackageJsonFile(project, extension)
              val parsedPackageJson = packageJson?.let { JsonUtils.fromCodegenJson(it) }

              val jsSrcsDirInPackageJson = parsedPackageJson?.codegenConfig?.jsSrcsDir
              if (jsSrcsDirInPackageJson != null) {
                it.jsRootDir.set(File(packageJson.parentFile, jsSrcsDirInPackageJson))
              } else {
                it.jsRootDir.set(extension.jsRootDir)
              }
              val needsCodegenFromPackageJson = project.needsCodegenFromPackageJson(extension)
              it.onlyIf { isLibrary || needsCodegenFromPackageJson }
            }

    // We create the task to generate Java code from schema.
    val generateCodegenArtifactsTask =
        project.tasks.register(
            "generateCodegenArtifactsFromSchema", GenerateCodegenArtifactsTask::class.java) {
              it.dependsOn(generateCodegenSchemaTask)
              it.reactNativeDir.set(extension.reactNativeDir)
              it.nodeExecutableAndArgs.set(extension.nodeExecutableAndArgs)
              it.generatedSrcDir.set(generatedSrcDir)
              it.packageJsonFile.set(findPackageJsonFile(project, extension))
              it.codegenJavaPackageName.set(extension.codegenJavaPackageName)
              it.libraryName.set(extension.libraryName)

              // Please note that appNeedsCodegen is triggering a read of the package.json at
              // configuration time as we need to feed the onlyIf condition of this task.
              // Therefore, the appNeedsCodegen needs to be invoked inside this lambda.
              val needsCodegenFromPackageJson = project.needsCodegenFromPackageJson(extension)
              it.onlyIf { isLibrary || needsCodegenFromPackageJson }
            }

    // We update the android configuration to include the generated sources.
    // This equivalent to this DSL:
    //
    // android { sourceSets { main { java { srcDirs += "$generatedSrcDir/java" } } } }
    project.extensions.getByType(AndroidComponentsExtension::class.java).finalizeDsl { ext ->
      ext.sourceSets.getByName("main").java.srcDir(File(generatedSrcDir, "java"))
    }

    // `preBuild` is one of the base tasks automatically registered by AGP.
    // This will invoke the codegen before compiling the entire project.
    project.tasks.named("preBuild", Task::class.java).dependsOn(generateCodegenArtifactsTask)
  }
}
