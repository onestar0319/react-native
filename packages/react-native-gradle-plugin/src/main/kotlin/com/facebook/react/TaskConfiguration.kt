/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react

import com.android.build.gradle.api.ApplicationVariant
import com.android.build.gradle.api.BaseVariant
import com.android.build.gradle.api.LibraryVariant
import com.android.build.gradle.internal.tasks.factory.dependsOn
import com.facebook.react.tasks.BundleJsAndAssetsTask
import com.facebook.react.tasks.HermesBinaryTask
import org.gradle.api.Project
import org.gradle.api.tasks.Copy
import org.gradle.kotlin.dsl.extra
import org.gradle.kotlin.dsl.register
import java.io.File

private const val REACT_GROUP = "react"

@Suppress("SpreadOperator")
internal fun Project.configureReactTasks(variant: BaseVariant, config: ReactAppExtension) {
  val targetName = variant.name.capitalize()
  val isRelease = variant.isRelease
  val targetPath = variant.dirName

  // React js bundle directories
  val jsBundleDir = File(buildDir, "generated/assets/react/$targetPath")
  val resourcesDir = File(buildDir, "generated/res/react/$targetPath")

  val jsBundleFile = File(jsBundleDir, config.bundleAssetName)
  val jsSourceMapsDir = File(buildDir, "generated/sourcemaps/react/$targetPath")
  val jsIntermediateSourceMapsDir = File(buildDir, "intermediates/sourcemaps/react/$targetPath")
  val jsPackagerSourceMapFile = File(jsIntermediateSourceMapsDir, "${config.bundleAssetName}.packager.map")
  val jsCompilerSourceMapFile = File(jsIntermediateSourceMapsDir, "${config.bundleAssetName}.compiler.map")
  val jsOutputSourceMapFile = File(jsSourceMapsDir, "${config.bundleAssetName}.map")

  // Additional node and packager commandline arguments
  val nodeExecutableAndArgs = config.nodeExecutableAndArgs
  val cliPath = config.detectedCliPath

  val execCommand = nodeExecutableAndArgs + cliPath
  val enableHermes = config.enableHermesForVariant(variant)
  val bundleEnabled = variant.checkBundleEnabled(config)

  val bundleTask = tasks.register<BundleJsAndAssetsTask>("createBundle${targetName}JsAndAssets") {
    val task = this
    task.group = REACT_GROUP
    task.description = "create JS bundle and assets for $targetName."

    task.reactRoot = config.reactRoot
    task.sources = fileTree(config.reactRoot) {
      setExcludes(config.inputExcludes)
    }
    task.execCommand = execCommand
    task.bundleCommand = config.bundleCommand
    task.devEnabled = !(variant.name in config.devDisabledInVariants || isRelease)
    task.entryFile = config.detectedEntryFile

    val extraArgs = mutableListOf<String>()

    if (config.bundleConfig != null) {
      extraArgs.add("--config")
      extraArgs.add(config.bundleConfig.orEmpty())
    }

    // Hermes doesn't require JS minification.
    if (enableHermes && !devEnabled) {
      extraArgs.add("--minify")
      extraArgs.add("false")
    }

    extraArgs.addAll(config.extraPackagerArgs)

    task.extraArgs = emptyList()

    task.jsBundleDir = jsBundleDir
    task.jsBundleFile = jsBundleFile
    task.resourcesDir = resourcesDir
    task.jsIntermediateSourceMapsDir = jsIntermediateSourceMapsDir
    task.jsSourceMapsDir = jsSourceMapsDir
    task.jsSourceMapsFile = if (enableHermes) jsPackagerSourceMapFile else jsOutputSourceMapFile

    task.enabled = bundleEnabled
  }

  val hermesTask = tasks.register<HermesBinaryTask>("emit${targetName}HermesResources") {
    val task = this
    task.group = REACT_GROUP
    task.description = "bundle hermes resources for $targetName"

    task.reactRoot = config.reactRoot
    task.hermesCommand = config.osAwareHermesCommand
    task.hermesFlags = if (isRelease) config.hermesFlagsRelease else config.hermesFlagsDebug
    task.jsBundleFile = jsBundleFile
    task.composeSourceMapsCommand = nodeExecutableAndArgs + config.composeSourceMapsPath
    task.jsPackagerSourceMapFile = jsPackagerSourceMapFile
    task.jsCompilerSourceMapFile = jsCompilerSourceMapFile
    task.jsOutputSourceMapFile = jsOutputSourceMapFile

    task.dependsOn(bundleTask)

    task.enabled = bundleEnabled && enableHermes
  }

  val aggregatedBundleTask = tasks.register("bundle${targetName}JsAndAssets") {
    val task = this
    task.group = REACT_GROUP
    task.description = "bundle JS and resources for $targetName"

    task.dependsOn(bundleTask, hermesTask)

    // this was exposed before, do we still need it?
    task.extra["generatedResFolders"] = files(resourcesDir).builtBy(task)
    task.extra["generatedAssetsFolders"] = files(jsBundleDir).builtBy(task)
  }

  val generatedResFolders = files(resourcesDir).builtBy(aggregatedBundleTask)

  // Android configuration
  variant.registerGeneratedResFolders(generatedResFolders)

  val packageTask = when (variant) {
    is ApplicationVariant -> variant.packageApplicationProvider
    is LibraryVariant -> variant.packageLibraryProvider
    else -> tasks.named("package$targetName")
  }

  val mergeResourcesTask = variant.mergeResourcesProvider
  val mergeAssetsTask = variant.mergeAssetsProvider
  val preBundleTask = tasks.named("build${targetName}PreBundle")

  val resourcesDirConfigValue = config.resourcesDir[variant.name]
  if (resourcesDirConfigValue != null) {
    val currentCopyResTask = tasks.register<Copy>("copy${targetName}BundledResources") {
      group = "react"
      description = "copy bundled resources into custom location for $targetName."

      from(resourcesDir)
      into(file(resourcesDirConfigValue))

      dependsOn(bundleTask)

      enabled = bundleEnabled
    }

    packageTask.dependsOn(currentCopyResTask)
    preBundleTask.dependsOn(currentCopyResTask)
  }

  packageTask.configure {
    if (config.enableVmCleanup) {
      doFirst {
        cleanupVMFiles(enableHermes, isRelease, targetPath)
      }
    }
  }

  val currentAssetsCopyTask = tasks.register<Copy>("copy${targetName}BundledJs") {
    group = "react"
    description = "copy bundled JS into $targetName."

    from(jsBundleDir)

    val jsBundleDirConfigValue = config.jsBundleDir[targetName]
    if (jsBundleDirConfigValue != null) {
      into(jsBundleDirConfigValue)
    } else {
      into(mergeAssetsTask.map { it.outputDir.get() })
    }

    dependsOn(mergeAssetsTask)

    enabled = bundleEnabled
  }

  // mergeResources task runs before the bundle file is copied to the intermediate asset directory from Android plugin 4.1+.
  // This ensures to copy the bundle file before mergeResources task starts
  mergeResourcesTask.dependsOn(currentAssetsCopyTask)
  packageTask.dependsOn(currentAssetsCopyTask)
  preBundleTask.dependsOn(currentAssetsCopyTask)
}

private fun Project.cleanupVMFiles(enableHermes: Boolean, isRelease: Boolean, targetPath: String) {
  // Delete the VM related libraries that this build doesn't need.
  // The application can manage this manually by setting 'enableVmCleanup: false'
  //
  // This should really be done by packaging all Hermes related libs into
  // two separate HermesDebug and HermesRelease AARs, but until then we'll
  // kludge it by deleting the .so files out of the /transforms/ directory.
  val libDir = "$buildDir/intermediates/transforms/"
  fileTree(libDir) {
    if (enableHermes) {
      // For Hermes, delete all the libjsc* files
      include("**/libjsc*.so")

      if (isRelease) {
        // Reduce size by deleting the debugger/inspector
        include("**/libhermes-inspector.so")
        include("**/libhermes-executor-debug.so")
      } else {
        // Release libs take precedence and must be removed
        // to allow debugging
        include("**/libhermes-executor-release.so")
      }
    } else {
      // For JSC, delete all the libhermes* files
      include("**/libhermes*.so")
    }
  }.visit {
    val targetVariant = ".*/transforms/[^/]*/$targetPath/.*".toRegex()
    val path = file.absolutePath.replace(File.separatorChar, '/')
    if (path.matches(targetVariant) && file.isFile()) {
      file.delete()
    }
  }
}

private fun BaseVariant.checkBundleEnabled(config: ReactAppExtension): Boolean {
  if (name in config.bundleIn) {
    return config.bundleIn.getValue(name)
  }

  if (buildType.name in config.bundleIn) {
    return config.bundleIn.getValue(buildType.name)
  }

  return isRelease
}

private val BaseVariant.isRelease: Boolean
  get() = name.toLowerCase().contains("release")
