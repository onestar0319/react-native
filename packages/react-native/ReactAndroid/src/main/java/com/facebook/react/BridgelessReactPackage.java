/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react;

import com.facebook.infer.annotation.Nullsafe;
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.devsupport.LogBoxModule;
import com.facebook.react.devsupport.interfaces.DevSupportManager;
import com.facebook.react.module.annotations.ReactModuleList;
import com.facebook.react.module.model.ReactModuleInfoProvider;
import com.facebook.react.modules.core.DefaultHardwareBackBtnHandler;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.facebook.react.modules.debug.DevSettingsModule;
import com.facebook.react.modules.debug.SourceCodeModule;
import com.facebook.react.modules.deviceinfo.DeviceInfoModule;
import com.facebook.react.modules.systeminfo.AndroidInfoModule;

@Nullsafe(Nullsafe.Mode.LOCAL)
@ReactModuleList(
    nativeModules = {
      AndroidInfoModule.class,
      DeviceInfoModule.class,
      DevSettingsModule.class,
      SourceCodeModule.class,
      LogBoxModule.class,
      DeviceEventManagerModule.class,
    })
public class BridgelessReactPackage extends TurboReactPackage {

  private DevSupportManager mDevSupportManager;
  private DefaultHardwareBackBtnHandler mHardwareBackBtnHandler;

  public BridgelessReactPackage(
      DevSupportManager devSupportManager, DefaultHardwareBackBtnHandler hardwareBackBtnHandler) {
    mDevSupportManager = devSupportManager;
    mHardwareBackBtnHandler = hardwareBackBtnHandler;
  }

  @Override
  public NativeModule getModule(String name, ReactApplicationContext reactContext) {
    switch (name) {
      case AndroidInfoModule.NAME:
        return new AndroidInfoModule(reactContext);
      case DeviceInfoModule.NAME:
        return new DeviceInfoModule(reactContext);
      case SourceCodeModule.NAME:
        return new SourceCodeModule(reactContext);
      case DevSettingsModule.NAME:
        return new DevSettingsModule(reactContext, mDevSupportManager);
      case DeviceEventManagerModule.NAME:
        return new DeviceEventManagerModule(reactContext, mHardwareBackBtnHandler);
      case LogBoxModule.NAME:
        return new LogBoxModule(reactContext, mDevSupportManager);
      default:
        throw new IllegalArgumentException(
            "In BridgelessReactPackage, could not find Native module for " + name);
    }
  }

  @Override
  public ReactModuleInfoProvider getReactModuleInfoProvider() {
    return new BridgelessReactPackage$$ReactModuleInfoProvider();
  }
}
