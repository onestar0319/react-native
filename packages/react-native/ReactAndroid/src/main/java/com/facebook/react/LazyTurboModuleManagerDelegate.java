/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react;

import androidx.annotation.Nullable;
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.common.annotations.UnstableReactNativeAPI;
import com.facebook.react.internal.turbomodule.core.interfaces.TurboModule;
import java.util.List;

/**
 * This abstract class provides a simple Lazy implementation of TurboModuleManagerDelegate. Main
 * difference between this class and ReactPackageTurboModuleManagerDelegate is that
 * LazyTurboModuleManagerDelegate does not require NativeModules to be annotated using a
 * ReactModule, also this class does not use the {@link
 * TurboReactPackage#getReactModuleInfoProvider} method. This class is for experimentation purposes
 * only, not to meant to be used in production.
 */
@UnstableReactNativeAPI
public abstract class LazyTurboModuleManagerDelegate
    extends ReactPackageTurboModuleManagerDelegate {

  private final List<ReactPackage> mPackages;
  private final ReactApplicationContext mReactContext;

  public LazyTurboModuleManagerDelegate(
      ReactApplicationContext reactApplicationContext, List<ReactPackage> packages) {
    super();
    mPackages = packages;
    mReactContext = reactApplicationContext;
  }

  @Override
  @Nullable
  public TurboModule getModule(String moduleName) {
    /*
     * Returns first TurboModule found with the name received as a parameter. There's no
     * warning or error if there are more than one TurboModule registered with the same name in
     * different packages. This method relies on the order of insertion of ReactPackage into
     * mPackages. Usually the size of mPackages is very small (2 or 3 packages in the majority of
     * the cases)
     */
    for (ReactPackage reactPackage : mPackages) {
      if (reactPackage instanceof BaseReactPackage) {
        BaseReactPackage baseReactPackage = (BaseReactPackage) reactPackage;
        try {
          TurboModule nativeModule =
              (TurboModule) baseReactPackage.getModule(moduleName, mReactContext);
          if (nativeModule != null) {
            return nativeModule;
          }
        } catch (IllegalArgumentException ex) {
          /*
           TurboReactPackages can throw an IllegalArgumentException when a module isn't found. If
           this happens, it's safe to ignore the exception because a later TurboReactPackage could
           provide the module.
          */
        }
      } else {
        throw new IllegalArgumentException("ReactPackage must be an instance of TurboReactPackage");
      }
    }
    return null;
  }

  @Override
  public boolean unstable_isModuleRegistered(String moduleName) {
    throw new UnsupportedOperationException("unstable_isModuleRegistered is not supported");
  }

  @Override
  public boolean unstable_isLazyTurboModuleDelegate() {
    return true;
  }

  @Override
  public boolean unstable_shouldEnableLegacyModuleInterop() {
    return false;
  }

  @Override
  public boolean unstable_shouldRouteTurboModulesThroughLegacyModuleInterop() {
    return false;
  }

  @Override
  public NativeModule getLegacyModule(String moduleName) {
    throw new UnsupportedOperationException("Legacy Modules are not supported");
  }

  @Override
  public boolean unstable_isLegacyModuleRegistered(String moduleName) {
    return false;
  };
}
