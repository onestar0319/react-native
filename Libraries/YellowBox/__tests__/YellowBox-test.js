/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+react_native
 * @format
 * @flow
 */

'use strict';

import * as React from 'react';
const YellowBox = require('../YellowBox');
const YellowBoxRegistry = require('../Data/YellowBoxRegistry');
const LogBoxData = require('../../LogBox/Data/LogBoxData');
const render = require('../../../jest/renderer');

jest.mock('../../LogBox/LogBoxNotificationContainer', () => ({
  __esModule: true,
  default: 'LogBoxNotificationContainer',
}));

describe('YellowBox', () => {
  const {error, warn} = console;

  beforeEach(() => {
    jest.resetModules();
    (console: any).error = jest.fn();
    (console: any).warn = jest.fn();
  });

  afterEach(() => {
    YellowBox.uninstall();
    (console: any).error = error;
    (console: any).warn = warn;
  });

  it('can set `disableYellowBox` after installing', () => {
    expect((console: any).disableYellowBox).toBe(undefined);

    YellowBox.install();

    expect((console: any).disableYellowBox).toBe(false);
    expect(YellowBoxRegistry.isDisabled()).toBe(false);

    (console: any).disableYellowBox = true;

    expect((console: any).disableYellowBox).toBe(true);
    expect(YellowBoxRegistry.isDisabled()).toBe(true);
  });

  it('can set `disableYellowBox` before installing', () => {
    expect((console: any).disableYellowBox).toBe(undefined);

    (console: any).disableYellowBox = true;
    YellowBox.install();

    expect((console: any).disableYellowBox).toBe(true);
    expect(YellowBoxRegistry.isDisabled()).toBe(true);
  });

  it('registers warnings', () => {
    jest.mock('../Data/YellowBoxRegistry');

    YellowBox.install();

    expect(YellowBoxRegistry.add).not.toBeCalled();
    (console: any).warn('...');
    expect(YellowBoxRegistry.add).toBeCalled();
  });

  it('registers errors beginning with "Warning: "', () => {
    jest.mock('../Data/YellowBoxRegistry');

    YellowBox.install();

    (console: any).error('...');
    expect(YellowBoxRegistry.add).not.toBeCalled();

    (console: any).error('Warning: ...');
    expect(YellowBoxRegistry.add).toBeCalled();
  });

  it('if LogBox is enabled, installs and uninstalls LogBox', () => {
    jest.mock('../../LogBox/Data/LogBoxData');
    jest.mock('../Data/YellowBoxRegistry');

    YellowBox.__unstable_enableLogBox();
    YellowBox.install();

    (console: any).warn('Some warning');
    expect(YellowBoxRegistry.add).not.toBeCalled();
    expect(LogBoxData.addLog).toBeCalled();
    expect(YellowBox.__unstable_isLogBoxEnabled()).toBe(true);

    YellowBox.uninstall();
    (LogBoxData.addLog: any).mockClear();

    (console: any).warn('Some warning');
    expect(YellowBoxRegistry.add).not.toBeCalled();
    expect(LogBoxData.addLog).not.toBeCalled();
    expect(YellowBox.__unstable_isLogBoxEnabled()).toBe(true);
  });

  it('throws if LogBox is enabled after YellowBox is installed', () => {
    jest.mock('../Data/YellowBoxRegistry');

    YellowBox.install();

    expect(() => YellowBox.__unstable_enableLogBox()).toThrow(
      'LogBox must be enabled before AppContainer is required so that it can properly wrap the console methods.\n\nPlease enable LogBox earlier in your app.\n\n',
    );
  });

  it('should render YellowBoxContainer by default', () => {
    const output = render.shallowRender(<YellowBox />);

    expect(output).toMatchSnapshot();
  });

  it('should render LogBoxNotificationContainer when LogBox is enabled', () => {
    YellowBox.__unstable_enableLogBox();

    const output = render.shallowRender(<YellowBox />);

    expect(output).toMatchSnapshot();
  });
});
