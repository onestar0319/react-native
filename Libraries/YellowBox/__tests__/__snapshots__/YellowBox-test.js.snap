// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YellowBox should render LogBoxNotificationContainer when LogBox is enabled 1`] = `<LogBoxNotificationContainer />`;

exports[`YellowBox should render YellowBoxContainer by default 1`] = `<YellowBoxContainer />`;
